packages/search-common/src/types.ts

```diff
@@ -53,6 +55,17 @@ export interface IndexableDocument {
    * is clicked).
    */
   location: string;
+
+  /**
+   * Optional authorization information to be used when determining whether this
+   * search result should be visible to a given user.
+   */
+  authorization?: {
```

Comment:
> I'm also realizing we're gonna collide in #7239 and regret not getting that in sooner.

😬 hopefully it's not too painful - are you expecting that to land imminently?

Reply:
I think you've got more momentum than I do; I'll probably come back in and just re-write once this is in.

Reply:
> On the topic of `(Perhaps) filter out resourceRef when returning results to the frontend.`
> 
> `IndexableDocument` does end up getting used/relied on in a variety of front-end contexts. Would this ever have any use in the front-end?

I don't think it's of use in the front end in it's current form, primarily because it's not guaranteed to be set. We might consider replacing the optional `authorization.resourceRef` with a required `resourceRef` field, and then we could expect it to be useful in contexts other than permissions, but I think we might have some more questions about how general purpose the resource ref is if we did.

> Might be nice to keep this out of the result interface and make it more of an implementation detail of the `SearchEngine.index` interface. Like...
> 
> ```ts
> type AuthorizableIndexableDocument = IndexableDocument & {
>   authorization?: { resourceRef: string; };
> };
> 
> interface SearchEngine {
>   index(type: string, documents: AuthorizableIndexableDocument[]): Promise<void>;
> }
> ```

I see what you're saying, but I'm left a little concerned that splitting this type up will be confusing, making backend consumers feel like they need to choose between the two, when they don't - they should always use `AuthorizableIndexableDocument`, since the `authorization` property is optional but should always have the right shape if it's set. 

I'd note that the `Indexable` part of the type name is also not super relevant to the frontend - what about introducing a new type for frontend use with a more generic name? Something like `SearchDocument = Omit<IndexableDocument, 'authorization'>` perhaps?

Reply:
> _...making backend consumers feel like they need to choose between the two, when they don't - they should always use `AuthorizableIndexableDocument`_

Good point.  I suppose you'd want this safety equally in decorators as in collators as in the search engine indexer.

Idle thought: would there be any value in making `authorization` more of a top-of-mind concern by actually requiring `authorization` on `IndexableDocument` with some other mechanism for indicating "no authorization needed, thanks" (e.g. `resourceRef` is required, but can be an empty string)?  I guess...  Imagine someone writing/contributing another collator.  Should authorization be a totally optional, opt-in thing for the collator author?  Or should they have to make an active decision?

> _I'd note that the `Indexable` part of the type name is also not super relevant to the frontend - what about introducing a new type for frontend use with a more generic name? Something like `SearchDocument = Omit<IndexableDocument, 'authorization'>` perhaps?_

I like this.  ...And carrying it through to the `SearchResult` / result set object.  ...Possibly making it generic in the future (to help folks not have to double-cast `someResult as unknown as CustomIndexableDocument` for type safety in custom result types)

Reply:
> Idle thought: would there be any value in making authorization more of a top-of-mind concern by actually requiring authorization on IndexableDocument with some other mechanism for indicating "no authorization needed, thanks" (e.g. resourceRef is required, but can be an empty string)? I guess... Imagine someone writing/contributing another collator. Should authorization be a totally optional, opt-in thing for the collator author? Or should they have to make an active decision?

Mmm - I think it's perhaps partially down to our appetite for breaking changes. I feel tempted (particularly given how long this PR's been open) to keep it optional for now and iterate, perhaps making it mandatory via a breaking change in the future.

The "explicit opt-out" option is also interesting, and seems quite closely related. It might well be that once the framework is established, we determine that it really should be the case that all documents in all collators should support permissions, in order to give integrators full control over visibility.

>  I like this. ...And carrying it through to the SearchResult / result set object. ...Possibly making it generic in the future (to help folks not have to double-cast someResult as unknown as CustomIndexableDocument for type safety in custom result types)

Sounds good! I'll make the initial change in a separate PR and we can go from there.

Reply:
`SearchDocument` added in https://github.com/backstage/backstage/pull/9202!